src/ray/gcs/gcs_server/gcs_server.cc

```diff
@@ -403,24 +400,38 @@ void GcsServer::StoreGcsServerAddressInRedis() {
   RAY_LOG(INFO) << "Finished setting gcs server address: " << address;
 }
 
-void GcsServer::InitResourceReportPolling(const GcsInitData &gcs_init_data) {
-  gcs_resource_report_poller_.reset(new GcsResourceReportPoller(
-      raylet_client_pool_, [this](const rpc::ResourcesData &report) {
-        gcs_resource_manager_->UpdateFromResourceReport(report);
-      }));
+void GcsServer::InitRaySyncer(const GcsInitData &gcs_init_data) {
+  /*
+    The current synchronization flow is:
+        syncer::poller -> gcs_resource_manager -> syncer::buffer -> syncer::broadcast
+
+    Ideally, we should make it:
+        syncer::poller --> gcs_resource_manager
+                       |-> syncer::buffer -> syncer::broadcast
+    But right now, placement group resource update is preventing us from doing this.
+    TODO (iycheng): Update placement group resource reporting and update it to the
+    right flow.
+  */
+  auto on_update = [this](const rpc::ResourcesData &report) {
+    main_service_.dispatch(
+        [this, report]() mutable {
+          gcs_resource_manager_->UpdateFromResourceReport(std::move(report));
+        },
+        "UpdateResourceReport");
+  };
 
-  gcs_resource_report_poller_->Initialize(gcs_init_data);
-  gcs_resource_report_poller_->Start();
-}
+  auto gcs_resource_report_poller = std::make_unique<GcsResourceReportPoller>(
+      raylet_client_pool_, std::move(on_update));
 
-void GcsServer::InitResourceReportBroadcasting(const GcsInitData &gcs_init_data) {
-  grpc_based_resource_broadcaster_ = std::make_unique<GrpcBasedResourceBroadcaster>(
-      raylet_client_pool_, [this](rpc::ResourceUsageBroadcastData &buffer) {
-        gcs_resource_manager_->GetResourceUsageBatchForBroadcast(buffer);
-      });
+  gcs_resource_report_poller->Initialize(gcs_init_data);
```

Comment:
Do we need separate Initialize API? Why can't this be done within a constructor?

Reply:
Because of pg code path, we can't put them into syncer constructor. It'll be removed in later PR.